src/passes/Outlining.cpp

```diff
@@ -13,23 +13,225 @@
+  //
+  // to know which IRBuilder to send the instruction.
+  enum ReconstructState {
+    NotInSeq = 0,
+    InSeq = 1,
+    InSkipSeq = 2,
+  };
+  ReconstructState state = ReconstructState::NotInSeq;
+  uint32_t seqCounter = 0;
+  uint32_t instrCounter = 0;
+  std::vector<OutliningSequence> sequences;
+  std::unordered_map<Name, std::vector<OutliningSequence>> seqToFunc;
+  IRBuilder existingBuilder;
+  IRBuilder outlinedBuilder;
+
+  void addUniqueSymbol(SeparatorReason reason) {
+    if (auto curr = reason.getFuncStart()) {
+      startExistingFunction(curr->func);
+      return;
+    }
+
+    // instrCounter is managed manually and incremented at the beginning of
+    // addUniqueSymbol() and visitExpression(), except for the case where we are
+    // starting a new function, as that resets the counters back to 0.
+    instrCounter++;
+
+    DBG(std::string desc);
+    if (auto curr = reason.getBlockStart()) {
+      ASSERT_ERR(existingBuilder.visitBlockStart(curr->block));
+      DBG(desc = "Block Start for ";);
+    } else if (auto curr = reason.getIfStart()) {
+      ASSERT_ERR(existingBuilder.visitIfStart(curr->iff));
+      DBG(desc = "If Start for ";);
+    } else if (reason.getEnd()) {
+      ASSERT_ERR(existingBuilder.visitEnd());
+      DBG(desc = "End for ";);
+    } else {
+      DBG(desc = "addUniqueSymbol for unimplemented control flow ";);
+      WASM_UNREACHABLE("unimplemented control flow");
+    }
+    DBG(printAddUniqueSymbol(desc););
+  }
+  void visitExpression(Expression* curr) {
+    maybeBeginSeq();
+
+    IRBuilder* builder = state == InSeq      ? &outlinedBuilder
+                         : state == NotInSeq ? &existingBuilder
+                                             : nullptr;
+    if (builder) {
+      ASSERT_ERR(builder->visit(curr));
+    }
+    DBG(printVisitExpression(curr));
+
+    if (state == InSeq || state == InSkipSeq) {
+      maybeEndSeq();
+    }
+  }
+  // Helpers
+  void startExistingFunction(Function* func) {
+    ASSERT_ERR(existingBuilder.build());
+    ASSERT_ERR(existingBuilder.visitFunctionStart(func));
+    instrCounter = 0;
+    seqCounter = 0;
+    state = NotInSeq;
+    DBG(std::cout << "\n\n$" << func->name << " Func Start "
+                  << &existingBuilder);
+  }
+  ReconstructState getCurrState() {
+    // We are either in a sequence or not in a sequence. If we are in a sequence
+    // and have already created the body of the outlined function that will be
+    // called, then we will skip instructions, otherwise we add the instructions
+    // to the outlined function. If we are not in a sequence, then the
+    // instructions are sent to the existing function.
+    if (seqCounter < sequences.size() &&
+        instrCounter >= sequences[seqCounter].startIdx &&
+        instrCounter < sequences[seqCounter].endIdx) {
+      return getModule()->getFunction(sequences[seqCounter].func)->body
+               ? InSkipSeq
+               : InSeq;
+    }
+    return NotInSeq;
```

Comment:
Just realized that it would be possible to replace this state enum with a single `IRBuilder* currBuilder` field that is either `&outlinedBuilder`, `&existingBuilder`, or `nullptr`. `getCurrentState` would simply set `currBuilder` to the correct value and we wouldn't need to worry about separately computing the right `IRBuilder` to use.

Up to you whether you want to do this, since the explicit state enum is slightly more self-documenting.

Reply:
Let's consider this a nice to have for a follow-up PR.